Remove course template from user page and user talk page when user disenrolls

gabina opened the pull request:
## What this PR does
The purpose of this PR is to add a way to remove course templates from user page and user talk page when user gets dis-enrolled from a specific course.

When a user enrolls in a course, a template pointing to the course is added to the user page and the user talk page (example user page with participating template: https://en.wikipedia.org/wiki/User:Jfitz19)

This PR adds a new `DisenrollFromCourseWorker` worker to remove course templates from user page and user talk page. It basically "reverts" what the `EnrollInCourseWorker` does (except for the user's sandbox action).

Now, when a user gets removed from a course, the `EnrollmentController#remove` schedules a job for the  `DisenrollFromCourseWorker` to remove the participating templates from user page and user talk page.

Closes #2842
## Screenshots

## Open questions and concerns

gabina commented on app/controllers/users/enrollment_controller.rb:
```diff
@@ -119,13 +119,19 @@ def remove
     update_course_page_and_assignment_talk_templates
   end
 
+  def remove_assignment_and_enrollment_templates
+    remove_assignment_templates
+    make_disenrollment_edits
+  end
```
Created this because got an "Method has too many lines" offense from rubocop

ragesoss replied:
I think a better way to reorganize this code would be to extract some of the setup and guard code at the start of the method. Ideally, the code in a method is at roughly the same level of abstraction, and the most detailed bits are the first ones to extract.

Extracting these two lines does not really make the method any easier to understand, which is the purpose behind that rubocop rule.

gabina replied:
Makes sense, I see your point. I tried to do that on 99392335e6128fb183c5a182d25d3c536deb9b7d

gabina commented on lib/wiki_course_edits.rb:
```diff
@@ -70,6 +70,25 @@ def enroll_in_course(enrolling_user:)
+    talk_page = "User_talk:#{disenrolling_user.username}"
+    talk_summary = "removing {{#{template_name(@templates, 'user_talk')}}}"
+    remove_content_from_page(talk_page, talk_template, talk_summary)
+  end
```
Sanity check: should I do something with the template in sandbox to revert the `add_template_to_sandbox` action?

ragesoss replied:
I don't think that's necessary, as that's not a place other editors will look to understand the context for the un-enrolled student's edits.

gabina replied:
I think I edited the message adding a second question after you answered, so just checking you saw both questions.

ragesoss replied:
Ah, I missed the second question. I think yes, checking `enrollment_edits_enabled?` for this action makes sense.

gabina replied:
Added the check and spec for that on 00cf010f45cfc3db2c60db59fc01c3e63189de8f

gabina commented:
Note: I was able to enable wiki edits on my local environment, and I reproduced this by adding and then removing my own user to a course. The history of how my user page and user talk page was modified can be check on https://en.wikipedia.org/w/index.php?title=User:Gabinaluz&action=history.

I didn't feel very confident reproducing that for fear of unintentional changes, so I didn't try more than that.

gabina commented:
CI failed with the following errors:
```
Finished in 31 minutes 9 seconds (files took 12.19 seconds to load)
1825 examples, 4 failures, 17 pending

Failed examples:

rspec ./spec/features/article_finder_spec.rb:17 # article finder performs searches and returns results
rspec ./spec/lib/wiki_pageviews_spec.rb:72 # WikiPageviews.views_for_article for an unviewed article returns an empty hash
rspec ./spec/lib/wiki_pageviews_spec.rb:151 # WikiPageviews.average_views_for_article for an article that does not exist returns 0
rspec ./spec/lib/wiki_pageviews_spec.rb:164 # WikiPageviews.average_views_for_article for an article that exist but has no view data returns 0
```
Will check if something is related to my changes.

ragesoss commented:
> Note: I was able to enable wiki edits on my local environment, and I reproduced this by adding and then removing my own user to a course. The history of how my user page and user talk page was modified can be check on https://en.wikipedia.org/w/index.php?title=User:Gabinaluz&action=history.
> 
> I didn't feel very confident reproducing that for fear of unintentional changes, so I didn't try more than that.

It looks like the edit summaries were buggy: `User has disenrolled in [[]]`. Do you know why that is?

gabina commented:
> It looks like the edit summaries were buggy: `User has disenrolled in [[]]`. Do you know why that is?

Oh good point.
The `disenrollment_summary` uses the `course.wiki_title` to build the summary, which needs the `wiki_course_page_enabled?` to return `true` to be other than `nil`. Due to the fact that on my local environment the `submitted` field for the course example was set to 0, `course.wiki_title` ended up being `nil`. I was playing around modifying the settings on my local environment and could get something like:
- User has enrolled in [Project:Outreach_Dashboard/Stanford_Law_School/Advanced_Legal_Research_Winter_2020_(Winter)](https://en.wikipedia.org/w/index.php?title=Wikipedia:Outreach_Dashboard/Stanford_Law_School/Advanced_Legal_Research_Winter_2020_(Winter)&action=edit&redlink=1).
- User has disenrolled in [Project:Outreach_Dashboard/Stanford_Law_School/Advanced_Legal_Research_Winter_2020_(Winter)](https://en.wikipedia.org/w/index.php?title=Wikipedia:Outreach_Dashboard/Stanford_Law_School/Advanced_Legal_Research_Winter_2020_(Winter)&action=edit&redlink=1).

You can check that on https://en.wikipedia.org/w/index.php?title=User:Gabinaluz&action=history
BTW, the site for the course doesn't exist (it was automatically created by my action, but I removed it). 

On the other hand, while taking a look at your point, I noticed a small bug in the disenrollment action, that wasn't caught by the tests. If you [compare the last two versions](https://en.wikipedia.org/w/index.php?title=User%3AGabinaluz&diff=1180488710&oldid=1180488539), you can notice that an extra new line is added when removing the course template, which is wrong. I modified the tests to catch the error, and then fixed it using a regular expression on 48d7a55c38dc2f4b8ac1355cd3c8897254a2110d

gabina commented on lib/wiki_course_edits.rb:
```diff
@@ -163,6 +182,16 @@ def add_template_to_sandbox
+    # If content to remove does not exist on initial page, then there is nothing
+    # to remove
+    return unless initial_page_content&.include?(content)
+    new_page_content = initial_page_content.gsub(/#{Regexp.quote(content)}(\n)?/, '')
```
I added an optional new line in the regexp, but maybe we can use just something like `#{Regexp.quote(content)}\n`, as the new line should be always added when writing the course template

ragesoss replied:
I guess the newline should be removed if present, but if there is not a newline, it should still remove the template.

gabina replied:
Makes sense.  Then it looks like using `/#{Regexp.quote(content)}(\n)?/` (as it's now) is the right way. If the new line after the content exists, it will match the content and the new line. If the new line doesn't exist, it will match only the content.

Let me know if there is something for me to review. Otherwise, I understand I already addressed all the suggestions/questions.

gabina commented:
I sent a request to join to wiki-education organization for Sentry because I cannot see the issue in Sentry

ragesoss commented:
@gabina that error is definitely not related to your change; I haven't deployed it to Programs & Events Dashboard yet.
